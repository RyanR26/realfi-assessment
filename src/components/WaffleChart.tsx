import React, { useMemo, useRef, Fragment } from 'react';
import counter from '../utils/counter'

interface IProps {
  dataPoints: IDataPoint[],
  config?: {
    title?: string,
    tooltipLabel?: string,
    height?: string,
    width?: string,
    containerPadding?: string
    groupClasses?: string[]
  }
}

interface IDataPoint { 
  value: number, 
  label: string
}

interface IData extends IDataPoint {
  percentage: number
}

export default function WaffleChart(props: IProps) {

  const { 
    dataPoints,
    config 
  } = props;

  var total = useMemo(() => dataPoints.reduce((accumulator, dataPoint: IDataPoint) => {
    return accumulator + dataPoint.value
  }, 0), [dataPoints]);

  const data = useMemo(() => dataPoints.map((dataPoint: IDataPoint) => ({
    ...dataPoint,
    percentage: Math.round((dataPoint.value/total) * 100)
  })), [dataPoints, total]);

  const countDivisor = config?.groupClasses?.length || 0;
  const counterRef = useRef(counter(countDivisor));
  const height = config?.height || '500px';
  const width = config?.width || '100%';

  return (
    <div className='waffle-chart'>
      { config?.title && <div className='text-md text-center pb-4 underline'>{ config.title }</div> }
      <div className='waffle-container rounded-box border-2 border-primary p-4 relative' style={{ height, width }}>
        <div className='grid grid-cols-10 w-full h-full gap-[1px]'>
        {
          data.map((dataPoint: IData, indexOuter: number) => {
            const colorIndex = counterRef.current(indexOuter);
            return (
            <Fragment key={indexOuter}>
              {
                [...Array(dataPoint.percentage).fill(dataPoint)].map((_, index) => (
                  <div 
                    key={index} 
                    className={`${config?.groupClasses ? config.groupClasses[colorIndex] : 'bg-secondary'} tooltip`}
                    style={{ borderRadius: '20%'}}
                    data-tip={`${dataPoint.label} - ${dataPoint.percentage}% ${config?.tooltipLabel}`}
                  >
                  </div>
                ))
              }            
            </Fragment> 
            )
          })  
        }
        </div>
      </div>
    </div>
  )
};