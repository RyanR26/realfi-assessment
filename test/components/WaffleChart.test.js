import renderer from 'react-test-renderer'
import WaffleChart from '../../src/components/WaffleChart'

it('Should render the ColChart HTML correctly against the snapshot', () => {

  const waffleChartDataPoints = [
    {
        "value": 2,
        "label": "Latvia"
    },
    {
        "value": 1,
        "label": "Spain"
    },
    {
        "value": 1,
        "label": "Germany"
    },
    {
        "value": 1,
        "label": "China"
    }
];

  const config = {
    "title": "User percentage by country",
    "tooltipLabel": "of users",
    "width": "300px",
    "height": "300px",
    "groupClasses": [
        "bg-primary",
        "bg-secondary",
        "bg-accent",
        "bg-neutral-content"
    ]
  };

  const component = renderer.create(<WaffleChart dataPoints={waffleChartDataPoints} config={config}/ >);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  {/* // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseEnter();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseLeave();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot(); */}
});