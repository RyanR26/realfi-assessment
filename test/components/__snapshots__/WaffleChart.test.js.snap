// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the ColChart HTML correctly against the snapshot 1`] = `
<div
  className="waffle-chart"
>
  <div
    className="text-md text-center pb-4 underline"
  >
    User percentage by country
  </div>
  <div
    className="waffle-container rounded-box border-2 border-primary p-4 relative"
    style={
      {
        "height": "300px",
        "width": "300px",
      }
    }
  >
    <div
      className="grid grid-cols-10 w-full h-full gap-[1px]"
    >
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-primary tooltip"
        data-tip="Latvia - 40% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-secondary tooltip"
        data-tip="Spain - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-accent tooltip"
        data-tip="Germany - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
      <div
        className="bg-neutral-content tooltip"
        data-tip="China - 20% of users"
        style={
          {
            "borderRadius": "20%",
          }
        }
      />
    </div>
  </div>
</div>
`;
